File: leetcode/0006_zigzag-conversion/solution.cc
#include <iostream>
#include <vector>
#include <algorithm> /* std::min */

class Solution {
public:
    std::string convert(std::string s, int numRows) {
        if (numRows == 1) return s;

        std::vector<std::string> vec(
            std::min( numRows,  int(s.size()) )
        );
        int row = 0;
        bool godown = false;
        for (char c: s) {
            vec[row] += c;
            if (row == 0 || row == numRows - 1)
                godown = !godown;
            row += godown ? 1 : -1;
        }

        std::string res;
        for (std::string v : vec) res += v;
        return res;
    }
};

/* 测试 */
int main(void)
{
    Solution *sln;
    std::string s = sln->convert("LEETCODEISHIRING", 3);
    std::cout << s << std::endl;
    delete sln;
    return 0;
}

File: leetcode/0006_zigzag-conversion/solution.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
  if (numRows === 1 || s.length < numRows) return s;

  // Solution 1
  // var result = '';
  // var n = (numRows - 1) * 2;
  // for (var i = 0; i < numRows; ++i) {
  //   var d = i * 2;
  //   for (var j = i; j < s.length; j += d) {
  //     result += s[j];
  //     d = n - d || n;
  //   }
  // }
  // return result;

  // Solution 2
  var result = [];
  var row = 0;
  var goDown = false;
  for (var c of s) {
    result[row] = (result[row] || '') + c;
    if (row == 0 || row == numRows - 1) goDown = !goDown;
    row += goDown ? 1 : -1;
  }
  return result.join('');
};

var data = [
  { s: 'AB', numRows: 1 },
  { s: 'ABC', numRows: 4 },
  { s: 'PAYPALISHIRING', numRows: 3 },
  { s: 'PAYPALISHIRING', numRows: 4 }
];
for (var i of data) {
  var result = convert(i.s, i.numRows);
  console.log(result);
}
// Output:
// AB
// ABC
// PAHNAPLSIIGYIR
// PINALSIGYAHRPI
